p2panda-sync/src/lib.rs

```diff
@@ -63,15 +146,63 @@ where
     ) -> Result<(), SyncError>;
 }
 
-#[derive(Debug, Error, PartialEq)]
+/// Messages which can be sent to the higher application layers (for further validation or
+/// persistance) and the underlying transport layer (for managing the sync session).
+#[derive(Debug, PartialEq)]
+pub enum FromSync<T>
+where
+    T: Topic,
+{
+    /// During the "Handshake" phase both peers usually manage access control and negotiate the
+    /// "topic" they want to exchange over. This message indicates that this phase has ended.
+    ///
+    /// Implementations for `p2panda-net` are required to send this message to the underlying
+    /// transport layer to inform the "backend" that we've successfully requested access, exchanged
+    /// the topic with the remote peer and are about to begin sync.
+    ///
+    /// With this information backends can optionally apply optimizations, which might for example
+    /// be required to keep application messages in-order (as there might exist other channels the
+    /// backend exchanges similar data over at the same time).
+    HandshakeSuccess(T),
+
+    /// Application data we've received during the sync session from the remote peer and want to
+    /// forward to higher application layers.
+    ///
+    /// These "frontends" might further process, decrypt payloads, sort messages or apply more
+    /// validation before they get finally persisted or rendered to the user. At this point the
+    /// sync protocol is merely "forwarding" it without any more knowledge how the data is used.
+    Data {
+        /// Exchanged data from sync session.
+        ///
+        /// Some data-types might be designed with "off-chain" use in mind, where a "header" is
+        /// crucial for integrity and authenticity but the actual "payload" is optional or
+        /// requested lazily in a later process.
+        header: Vec<u8>,
+
+        /// Optional "body" which can represent "off-chain" application data.
+        ///
+        /// This is useful for realizing "off-chain" compatible data types. Implementations without
+        /// this distinction will leave this field always to `None` and only encode their data
+        /// types in the `header` field.
+        payload: Option<Vec<u8>>,
+    },
+}
```

Comment:
Thought its nicer if `header` and `payload` have distinct named fields, this might make it easier to understand what one is looking at (instead of `Vec<u8>, Option<Vec<u8>>`. I don't like it introduces the word "header" (not everyone will think about it like that) but I think that's very edge-case.